sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubClient.cs

```diff
@@ -126,15 +160,30 @@ public EventHubClient(string connectionString) : this(connectionString, null)
                               TokenCredential credential,
                               EventHubClientOptions clientOptions = default)
         {
+            clientOptions = clientOptions?.Clone() ?? new EventHubClientOptions();
+
             Guard.ArgumentNotNullOrEmpty(nameof(host), host);
             Guard.ArgumentNotNullOrEmpty(nameof(eventHubPath), eventHubPath);
             Guard.ArgumentNotNull(nameof(credential), credential);
             ValidateClientOptions(clientOptions);
 
+            switch (credential)
+            {
+                case SharedAccessSignatureCredential _:
+                    break;
+
+                case EventHubSharedKeyCredential sharedKeyCredential:
+                    credential = sharedKeyCredential.ConvertToSharedAccessSignatureCredential(BuildResource(clientOptions.TransportType, host, eventHubPath));
+                    break;
+
+                default:
+                    credential = new EventHubTokenCredential(credential, BuildResource(clientOptions.TransportType, host, eventHubPath));
+                    break;
+            }
+
```

Comment:
My one concern with this approach is that it's not very explicit what the user can pass to this ctor.  While it is convenient and might be warranted due to the number of constructors I wonder if users will know the options that are available here.  I'm not saying specifically that we should change this, but I'd like to get your thoughts on it.   Also, The doc comment says "The Azure managed identity credential..." which will be confusing since there is a ManagedIdenityCredential class in Azure.Idenity.  If we don't change this to have specific ctors for each type of credential the client excepts we should explicitly list them in the doc comment.  Thoughts?

Reply:
I'm not sure that I'm following completely with respect to the user potentially not understanding the options available.   Are you suggesting that there are too many overall constructors or that there is something specific to this expanded one that may be a concern?   Or, are you thinking specifically in terms of the token accepted?

As far as the comment goes, I went with the best understanding that I had for what the identity library was exposing.  Since you've got a much better understanding there, I'd appreciate any suggestion for clarifying...  "an Azure Identity credential?"   Something else?

Reply:
I was talking about all the types which can be passed here.  The user could pass a SharedAccessSignatureCredential, a EventHubSharedKeyCredential, or any implementation of a TokenCredential, which could be their own implementation or any of the implementations available in Azure.Identity.  This is because you've derived SharedAccessSignatureCredential and EventHubSharedKeyCredential from TokenCredential rather than having explicit constructors which take those types, which is the pattern that we're trying use for other client libraries. 

That being said perhaps this is warranted, as I was saying above, since adding explicit constructors for each credential type would grow the total ctor overloads to 5, but we would need to document what can be passed to this constructor which will likely be fairly verbose.  Is that more clear?

Reply:
Thanks, Scott.   I'm following now.   The only non-internal type from within the Event Hubs library would be `EventHubSharedKeyCredential`,  the `SharedAccessSignatureType` is an internal construct used to unify the flow of logic.    

That said, it's an interesting question on how much exposure we want to offer to the shared key credential.   The primary use case that we've got now is for that to be taken from the connection string, though I did want to allow for it to be passed in as any other token credential could, though maybe that's overdoing it.   I added the type before making the decision to expand the constructors to work around difficulty discovering the right connection string in the portal.   I'm not sure that adding another two overloads is the way to go - though I'm not sure that removing the public credential is either.   Out of the two, I think that I'm leaning more towards removing the type.

I think I'm going to take a note, think on this, and punt it until after preview so that we can see feedback and make a more informed call, unless you feel more strongly about it, in which case I can remove it now.

Reply:
Sounds good.  Let's add an issue to track this.

---
In reply to: [294936718](https://github.com/Azure/azure-sdk-for-net/pull/6599#discussion_r294936718) [](ancestors = 294936718)

Reply:
Opened [#6618](https://github.com/Azure/azure-sdk-for-net/issues/6618) to track.